charms/k8s/src/charm.py

```diff
@@ -83,13 +84,30 @@ def _apply_snap_requirements(self):
         for c in commands:
             subprocess.check_call(shlex.split(c))
 
-    @on_error(ops.WaitingStatus("Failed to bootstrap k8s snap"), subprocess.CalledProcessError)
+    @on_error(WaitingStatus("Waiting for k8sd"), InvalidResponseError, K8sdConnectionError)
+    def _check_k8sd_ready(self):
+        """Check if k8sd is ready to accept requests."""
+        status.add(ops.MaintenanceStatus("Check k8sd ready"))
+        self.api_manager.check_k8sd_ready()
+
+    @on_error(
+        ops.WaitingStatus("Failed to bootstrap k8s snap"),
+        InvalidResponseError,
+        K8sdConnectionError,
+    )
     def _bootstrap_k8s_snap(self):
-        """Bootstrap the k8s if it's not already bootstrapped."""
+        """Bootstrap k8s if it's not already bootstrapped."""
+        # TODO: Remove `is_cluster_bootstrapped` check once
+        # https://github.com/canonical/k8s-snap/pull/99 landed.
         if not self.api_manager.is_cluster_bootstrapped():
             status.add(ops.MaintenanceStatus("Bootstrapping Cluster"))
-            cmd = "k8s bootstrap"
-            subprocess.check_call(shlex.split(cmd))
+            binding = self.model.get_binding("juju-info")
+            address = binding and binding.network.ingress_address
+            # k8s/x to k8s-x to avoid trouble with urls
+            name = self.unit.name.replace("/", "-")
+
+            # TODO: Make port (and address) configurable.
+            self.api_manager.bootstrap_k8s_snap(name, f"{str(address)}:6400")
```

Comment:
@mateoflorido shouldn't this bootstrapping really only occur on the charm leader?  It seems to me we can't bootstrap every control-plane node independently

Reply:
Clustering is not part of this PR. Let's add this check once we need it.

Reply:
👍🏻 agree, i see that on the clustering PR.  No worries